File: infra/runners/python/supervisor.py
#!/usr/bin/env python3
"""
Botforge Python bot runner — supervisor.

This is OUR code, not the user's. It runs as PID 1 inside the Fly Machine and is
never editable by the user or the AI. It:

  1. pulls the project's files from the Botforge control plane (by run token),
  2. writes them to /app  (never a .env — secrets arrive as real env vars),
  3. installs requirements.txt if present,
  4. execs the bot's entry file and streams its stdout/stderr back as logs,
  5. reports the process exit so the platform can mark the run crashed/stopped.

Only the Python standard library is used here; the bot's own dependencies are
installed at runtime from its requirements.txt.

Contract (env vars injected by the start route):
  BOTFORGE_PUBLIC_URL  base URL of the Botforge app (e.g. https://botforge.app)
  BOTFORGE_RUN_TOKEN   run-scoped bearer token (authenticates the callbacks)
  BOTFORGE_ENTRY       entry file to run, e.g. "main.py"
  <secret keys>        e.g. TELEGRAM_TOKEN — the bot reads these from os.environ
"""
import json
import os
import queue
import subprocess
import sys
import threading
import time
import urllib.error
import urllib.request

BASE = os.environ.get("BOTFORGE_PUBLIC_URL", "").rstrip("/")
TOKEN = os.environ.get("BOTFORGE_RUN_TOKEN", "")
ENTRY = os.environ.get("BOTFORGE_ENTRY", "main.py")
APP_DIR = "/app"

FLUSH_INTERVAL = 1.0      # seconds between log flushes
FLUSH_MAX_BATCH = 40      # or flush early once this many lines are queued
LINE_MAX = 2000           # truncate absurdly long lines


def _post(path: str, payload: dict, timeout: float = 15.0) -> bytes:
    req = urllib.request.Request(
        f"{BASE}{path}",
        data=json.dumps(payload).encode("utf-8"),
        headers={"Content-Type": "application/json", "Authorization": f"Bearer {TOKEN}"},
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=timeout) as r:
        return r.read()


def _get(path: str, timeout: float = 30.0) -> dict:
    req = urllib.request.Request(
        f"{BASE}{path}",
        headers={"Authorization": f"Bearer {TOKEN}"},
        method="GET",
    )
    with urllib.request.urlopen(req, timeout=timeout) as r:
        return json.loads(r.read().decode("utf-8"))


# --- log shipping ---------------------------------------------------------

_log_q: "queue.Queue[dict]" = queue.Queue(maxsize=10000)
_stop = threading.Event()


def enqueue(stream: str, line: str) -> None:
    try:
        _log_q.put_nowait({"stream": stream, "line": line[:LINE_MAX]})
    except queue.Full:
        pass  # drop under extreme volume rather than block the bot


def flusher() -> None:
    """Drain queued log lines to the control plane in small batches."""
    while not _stop.is_set() or not _log_q.empty():
        batch = []
        deadline = time.time() + FLUSH_INTERVAL
        while len(batch) < FLUSH_MAX_BATCH and time.time() < deadline:
            try:
                batch.append(_log_q.get(timeout=0.2))
            except queue.Empty:
                if batch:
                    break
        if not batch:
            continue
        try:
            _post("/api/internal/hosting/logs", {"lines": batch})
        except Exception:
            pass  # never let logging failures kill the bot


def pump(pipe, stream: str) -> None:
    for raw in iter(pipe.readline, ""):
        enqueue(stream, raw.rstrip("\n"))
    pipe.close()


# --- system log helper (goes to the "system" stream) ----------------------

def system(msg: str) -> None:
    try:
        _post("/api/internal/hosting/logs", {"lines": [{"stream": "system", "line": msg}]})
    except Exception:
        pass
    print(f"[botforge] {msg}", flush=True)


def report_exit(code: int) -> None:
    try:
        _post("/api/internal/hosting/exit", {"code": code})
    except Exception:
        pass


# --- setup ----------------------------------------------------------------

def write_files() -> None:
    data = _get("/api/internal/hosting/files")
    files = data.get("files", [])
    written = 0
    for f in files:
        path = f.get("path", "")
        content = f.get("content", "")
        if not path:
            continue
        # Never materialize a .env — secrets are injected as real env vars, and a
        # stray dotenv file must not shadow or blank them.
        base = path.split("/")[-1]
        if base == ".env" or base.startswith(".env."):
            continue
        # Contain writes to APP_DIR (defensive against path traversal).
        norm = os.path.normpath(os.path.join(APP_DIR, path))
        if not norm.startswith(APP_DIR + os.sep) and norm != APP_DIR:
            continue
        os.makedirs(os.path.dirname(norm), exist_ok=True)
        with open(norm, "w", encoding="utf-8") as out:
            out.write(content)
        written += 1
    system(f"Fetched {written} file(s).")


def install_requirements() -> None:
    req = os.path.join(APP_DIR, "requirements.txt")
    if not os.path.exists(req):
        return
    system("Installing requirements…")
    proc = subprocess.run(
        [sys.executable, "-m", "pip", "install", "--no-cache-dir", "--disable-pip-version-check", "-r", req],
        cwd=APP_DIR, capture_output=True, text=True,
    )
    for line in (proc.stdout or "").splitlines()[-40:]:
        enqueue("system", line)
    if proc.returncode != 0:
        for line in (proc.stderr or "").splitlines()[-40:]:
            enqueue("stderr", line)
        raise RuntimeError(f"pip install failed (exit {proc.returncode})")


def main() -> int:
    if not BASE or not TOKEN:
        print("[botforge] missing BOTFORGE_PUBLIC_URL / BOTFORGE_RUN_TOKEN", file=sys.stderr, flush=True)
        return 2

    flush_thread = threading.Thread(target=flusher, daemon=True)
    flush_thread.start()

    try:
        write_files()
        install_requirements()
    except Exception as e:
        system(f"Startup failed: {e}")
        _stop.set()
        flush_thread.join(timeout=10)
        report_exit(1)
        return 1

    entry_path = os.path.join(APP_DIR, ENTRY)
    if not os.path.exists(entry_path):
        system(f"Entry file '{ENTRY}' not found.")
        _stop.set()
        flush_thread.join(timeout=10)
        report_exit(1)
        return 1

    system(f"Starting: python {ENTRY}")
    proc = subprocess.Popen(
        [sys.executable, "-u", ENTRY],
        cwd=APP_DIR,
        stdout=subprocess.PIPE, stderr=subprocess.PIPE,
        text=True, bufsize=1,
        env={**os.environ, "PYTHONUNBUFFERED": "1"},
    )
    t_out = threading.Thread(target=pump, args=(proc.stdout, "stdout"), daemon=True)
    t_err = threading.Thread(target=pump, args=(proc.stderr, "stderr"), daemon=True)
    t_out.start()
    t_err.start()

    code = proc.wait()
    t_out.join(timeout=5)
    t_err.join(timeout=5)
    system(f"Process exited with code {code}.")

    _stop.set()
    flush_thread.join(timeout=10)
    report_exit(code)
    return code


if __name__ == "__main__":
    sys.exit(main())

File: infra/runners/node/supervisor.js
#!/usr/bin/env node
/**
 * Botforge Node bot runner — supervisor.
 *
 * This is OUR code, not the user's. It runs as PID 1 inside the Fly Machine and
 * is never editable by the user or the AI. Mirrors `../python/supervisor.py`'s
 * contract exactly (same env vars, same callback routes, same behavior) — this
 * is the Node/JS twin, used for Telegram bots written in Node (grammy) and all
 * Discord bots (discord.js). It:
 *
 *   1. pulls the project's files from the Botforge control plane (by run token),
 *   2. writes them to /app (never a .env — secrets arrive as real env vars),
 *   3. npm installs from package.json if present,
 *   4. execs the bot's entry file and streams its stdout/stderr back as logs,
 *   5. reports the process exit so the platform can mark the run crashed/stopped.
 *
 * Only Node's built-ins are used here (fetch, fs, path, child_process); the
 * bot's own dependencies are installed at runtime from its package.json.
 *
 * Contract (env vars injected by the start route):
 *   BOTFORGE_PUBLIC_URL  base URL of the Botforge app (e.g. https://botforge.app)
 *   BOTFORGE_RUN_TOKEN   run-scoped bearer token (authenticates the callbacks)
 *   BOTFORGE_ENTRY       entry file to run, e.g. "index.js"
 *   <secret keys>        e.g. DISCORD_TOKEN — the bot reads these from process.env
 */
"use strict";

const fs = require("fs");
const path = require("path");
const { spawn, spawnSync } = require("child_process");

const BASE = (process.env.BOTFORGE_PUBLIC_URL || "").replace(/\/$/, "");
const TOKEN = process.env.BOTFORGE_RUN_TOKEN || "";
const ENTRY = process.env.BOTFORGE_ENTRY || "index.js";
const APP_DIR = "/app";

const FLUSH_INTERVAL_MS = 1000; // pace between log flushes
const FLUSH_MAX_BATCH = 40; // cap lines per flush
const LINE_MAX = 2000; // truncate absurdly long lines
const LOG_QUEUE_MAX = 10000; // drop under extreme volume rather than block the bot

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function post(pathName, payload, timeoutMs = 15000) {
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), timeoutMs);
  try {
    await fetch(`${BASE}${pathName}`, {
      method: "POST",
      headers: { "Content-Type": "application/json", Authorization: `Bearer ${TOKEN}` },
      body: JSON.stringify(payload),
      signal: ctrl.signal,
    });
  } finally {
    clearTimeout(timer);
  }
}

// Retried: a transient network blip at boot must not become a false "crash".
async function get(pathName, { timeoutMs = 30000, attempts = 3 } = {}) {
  let lastErr;
  for (let i = 0; i < attempts; i++) {
    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), timeoutMs);
    try {
      const res = await fetch(`${BASE}${pathName}`, {
        headers: { Authorization: `Bearer ${TOKEN}` },
        signal: ctrl.signal,
      });
      if (!res.ok) throw new Error(`GET ${pathName} -> ${res.status}`);
      return await res.json();
    } catch (e) {
      lastErr = e;
      if (i < attempts - 1) await sleep(2000 * (i + 1));
    } finally {
      clearTimeout(timer);
    }
  }
  throw lastErr;
}

// --- log shipping -----------------------------------------------------------

const logQueue = [];
let stopping = false;

function enqueue(stream, line) {
  if (logQueue.length >= LOG_QUEUE_MAX) return;
  logQueue.push({ stream, line: String(line).slice(0, LINE_MAX) });
}

/** Drains queued log lines to the control plane in small batches. Runs until
 * `stopping` is set AND the queue is empty — callers await this to flush. */
async function flusher() {
  for (;;) {
    const batch = logQueue.splice(0, FLUSH_MAX_BATCH);
    if (batch.length) {
      try {
        await post("/api/internal/hosting/logs", { lines: batch });
      } catch {
        /* never let logging failures kill the bot */
      }
    }
    if (stopping && logQueue.length === 0) return;
    await sleep(FLUSH_INTERVAL_MS);
  }
}

/** Buffers partial chunks and enqueues one log line per newline. */
function pump(stream, streamName) {
  let buf = "";
  stream.on("data", (chunk) => {
    buf += chunk.toString("utf8");
    let idx;
    while ((idx = buf.indexOf("\n")) >= 0) {
      enqueue(streamName, buf.slice(0, idx));
      buf = buf.slice(idx + 1);
    }
  });
  stream.on("end", () => {
    if (buf) enqueue(streamName, buf);
  });
}

// --- system log helper (goes to the "system" stream) ------------------------

function system(msg) {
  post("/api/internal/hosting/logs", { lines: [{ stream: "system", line: msg }] }).catch(() => {});
  console.log(`[botforge] ${msg}`);
}

function reportExit(code) {
  return post("/api/internal/hosting/exit", { code }).catch(() => {});
}

// --- setup --------------------------------------------------------------

async function writeFiles() {
  const data = await get("/api/internal/hosting/files");
  const files = data.files || [];
  let written = 0;
  for (const f of files) {
    const filePath = f.path || "";
    const content = f.content || "";
    if (!filePath) continue;
    // Never materialize a .env — secrets are injected as real env vars, and a
    // stray dotenv file must not shadow or blank them.
    const base = filePath.split("/").pop();
    if (base === ".env" || base.startsWith(".env.")) continue;
    // Contain writes to APP_DIR (defensive against path traversal).
    const norm = path.normalize(path.join(APP_DIR, filePath));
    if (!norm.startsWith(APP_DIR + path.sep) && norm !== APP_DIR) continue;
    fs.mkdirSync(path.dirname(norm), { recursive: true });
    fs.writeFileSync(norm, content, "utf8");
    written += 1;
  }
  system(`Fetched ${written} file(s).`);
}

function installDependencies() {
  const pkgPath = path.join(APP_DIR, "package.json");
  if (!fs.existsSync(pkgPath)) return;
  system("Installing dependencies…");
  const result = spawnSync("npm", ["install", "--omit=dev", "--no-audit", "--no-fund"], {
    cwd: APP_DIR,
    encoding: "utf8",
  });
  for (const line of (result.stdout || "").split("\n").slice(-40)) if (line) enqueue("system", line);
  if (result.status !== 0) {
    for (const line of (result.stderr || "").split("\n").slice(-40)) if (line) enqueue("stderr", line);
    throw new Error(`npm install failed (exit ${result.status})`);
  }
}

async function main() {
  if (!BASE || !TOKEN) {
    console.error("[botforge] missing BOTFORGE_PUBLIC_URL / BOTFORGE_RUN_TOKEN");
    process.exit(2);
  }

  const flushDone = flusher();

  try {
    await writeFiles();
    installDependencies();
  } catch (e) {
    system(`Startup failed: ${e.message}`);
    stopping = true;
    await flushDone;
    await reportExit(1);
    process.exit(1);
  }

  const entryPath = path.join(APP_DIR, ENTRY);
  if (!fs.existsSync(entryPath)) {
    system(`Entry file '${ENTRY}' not found.`);
    stopping = true;
    await flushDone;
    await reportExit(1);
    process.exit(1);
  }

  system(`Starting: node ${ENTRY}`);
  // The bot only gets its own secrets. The platform's callback credentials
  // (BOTFORGE_RUN_TOKEN etc.) stay with the supervisor — user/AI code must
  // never be able to read or reuse them.
  const childEnv = {};
  for (const [k, v] of Object.entries(process.env)) {
    if (!k.startsWith("BOTFORGE_")) childEnv[k] = v;
  }

  const child = spawn(process.execPath, [ENTRY], {
    cwd: APP_DIR,
    env: childEnv,
    stdio: ["ignore", "pipe", "pipe"],
  });
  pump(child.stdout, "stdout");
  pump(child.stderr, "stderr");

  const code = await new Promise((resolve) => {
    child.on("close", (exitCode) => resolve(exitCode ?? 0));
  });
  system(`Process exited with code ${code}.`);

  stopping = true;
  await flushDone;
  await reportExit(code);
  process.exit(code);
}

main();
